solver/src/settlement_submission/flashbots_settlement.rs

```diff
@@ -0,0 +1,371 @@
+// find the one that got mined (if any).
+
+use super::{flashbots_api::FlashbotsApi, ESTIMATE_GAS_LIMIT_FACTOR};
+use crate::{interactions::block_coinbase, settlement::Settlement};
+use anyhow::{anyhow, ensure, Context, Result};
+use contracts::GPv2Settlement;
+use ethcontract::{errors::MethodError, transaction::Transaction, Account};
+use futures::FutureExt;
+use gas_estimation::{EstimatedGasPrice, GasPriceEstimating};
+use primitive_types::{H256, U256};
+use shared::Web3;
+use std::time::{Duration, Instant};
+use web3::types::TransactionReceipt;
+
+pub struct FlashbotsSolutionSubmitter<'a> {
+    web3: &'a Web3,
+    contract: &'a GPv2Settlement,
+    // Invariant: MUST be an `Account::Offline`.
+    account: &'a Account,
+    flashbots_api: &'a FlashbotsApi,
+    gas_price_estimator: &'a dyn GasPriceEstimating,
+    gas_price_cap: f64,
+}
+
+impl<'a> FlashbotsSolutionSubmitter<'a> {
+    pub fn new(
+        web3: &'a Web3,
+        contract: &'a GPv2Settlement,
+        account: &'a Account,
+        flashbots_api: &'a FlashbotsApi,
+        gas_price_estimator: &'a dyn GasPriceEstimating,
+        gas_price_cap: f64,
+    ) -> Result<Self> {
+        ensure!(
+            matches!(account, Account::Offline(..)),
+            "Flashbots submission requires offline account for signing"
+        );
+
+        Ok(Self {
+            web3,
+            contract,
+            account,
+            flashbots_api,
+            gas_price_estimator,
+            gas_price_cap,
+        })
+    }
+}
+
+impl<'a> FlashbotsSolutionSubmitter<'a> {
+    /// Submit a settlement to the contract, updating the transaction with gas prices if they increase.
+    ///
+    /// Goes through the flashbots network so that failing transactions do not get mined and thus do
+    /// not cost gas.
+    ///
+    /// Only works on mainnet.
+    pub async fn submit(
+        &self,
+        target_confirm_time: Duration,
+        settlement: Settlement,
+        gas_estimate: U256,
+    ) -> Result<Option<TransactionReceipt>> {
+        let nonce = self.nonce().await?;
+
+        tracing::info!("starting flashbots solution submission at nonce {}", nonce,);
+
+        let mut transactions = Vec::new();
+        let submit_future = self.submit_with_increasing_gas_prices_until_simulation_fails(
+            target_confirm_time,
+            nonce,
+            settlement,
+            gas_estimate,
+            &mut transactions,
+        );
+
+        let nonce_future = self.wait_for_nonce_to_change(nonce);
+
+        futures::select! {
+            method_error = submit_future.fuse() => tracing::info!("stopping submission because simulation failed: {:?}", method_error),
+            new_nonce = nonce_future.fuse() => tracing::info!("stopping submission because account nonce changed to {}", new_nonce),
+        };
+
+        // After stopping submission of new transactions we wait for some time to give a potentially
```

Comment:
Just curious why do we need to wait for the tx to show up on our node?

Reply:
Because there is a delay between submission of the tx and the receival of the mined block containing the tx.
I guess Archer API does not support tx status query, while Flashbots does. However, it seemed to me that it is better to rely on the network to check if tx is mined, rather then trust Flashbots API.